lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java

```diff
@@ -217,8 +250,42 @@ public void execute() throws IOException, RecommenderBuildException {
                 if (n > 0) {
                     rmse = sqrt(sse / n);
                 }
-                //writes the Prediction Score and TARMSE on file
-                tableWriter.writeRow(r.getUserId(), r.getItemId(), r.getValue(), r.getTimestamp(), predict, rmse);
+
+                //calculate recommendation rank
+                int rank = -1;
+                LongSet itemsInDao = new LongOpenHashSet(limitedDao.getItemIds().size());
+                itemsInDao.addAll(limitedDao.getItemIds());
+
+                if (itemsInDao.size() > 0) {
+                    //candidate =  r.getItemId() + random (items from dao - items by user, listsize-1);
+                    LongSet itemsByUser = new LongOpenHashSet(itemsInDao.size());
+
+                    if (limitedDao.getEventsForUser(r.getUserId()) != null) {
+                        itemsByUser.addAll(limitedDao.getEventsForUser(r.getUserId()).itemSet());
+
+                    }
+                    itemsInDao.removeAll(itemsByUser);
+                    LongSet candidates = new LongOpenHashSet(10);
+
+                    if (itemsInDao.size() > 0) {
+                        candidates.addAll(LongUtils.randomSubset(itemsInDao, listSize - 1, new Random()));
```

Comment:
Let's declare a field `Random rng` on the temporal evaluator, instead of creating a new `Random` every time.